Update versions and add more advanced content (#30)
actions-workshop/actions-workshop: docs/306-authoring-actions.md

**Javascript Actions** are custom actions written in JavaScript or Typescript that run directly on the runner.

**Docker Actions** are custom actions that run inside a Docker container.  This allows you full control over the runtime environment, but does add some overhead.

## Prerequisites
- Ideally you will complete the module on reusable workflows first, but it is not required.

---

## Part 1: Composite Workflows

Composite workflows allow you to define a workflow that can be called from other workflows.  This is useful for grouping common steps that are used in multiple workflows.  Unlike reusable workflows, composite workflows are defined as actions and are called as a single step in a job.

In this lab we will extract build and test steps from an existing workflow into a reusable composite workflow.

Take this simple workflow that builds, tests, and lints a Node.js app:

```yaml
# .github/workflows/ci.yml
name: CI

on:
  push:
    branches: [ main ]
  pull_request:
    branches: [ main ]
  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest
    steps:

>       - uses: actions/checkout@v4
>       - uses: actions/setup-node@v4
>         with:
>           node-version: '20'

Reviewer: ```suggestion
      - uses: actions/checkout@v5
      - uses: actions/setup-node@v5
        with:
          node-version: '24'
```

      - run: npm ci
      - run: npm test

  lint:
    runs-on: ubuntu-latest
    steps:

>       - uses: actions/checkout@v4
>       - uses: actions/setup-node@v4
>         with:
>           node-version: '20'

Reviewer: ```suggestion
      - uses: actions/checkout@v5
      - uses: actions/setup-node@v5
        with:
          node-version: '24'
```

      - run: npm run lint
```

Note the repeated steps for checking out the code, setting up Node.js, and installing dependencies.  We can extract these into a reusable composite workflow.

Create a new directory for the composite action:

```bash
mkdir -p .github/actions/setup-and-install
```

Create a file for the composite workflow at `.github/actions/setup-and-install/action.yml`.  Note the use of an action.yml file to define the composite action.  This aligns with how JavaScript and Docker actions are defined.  Use the following content:

```yaml
name: 'Setup and Install'
description: 'Checks out code, sets up Node, and installs dependencies'
inputs:
  node-version:
    description: 'Node.js version'
    required: true
runs:
  using: 'composite'
  steps:
    - uses: actions/setup-node@v4
      with:
        node-version: ${{ inputs.node-version }}
    - run: npm ci
      shell: bash
```

Now create a new workflow that uses this composite action.  Create `.github/workflows/composite-call-ci.yml` with the following content:

```yaml
name: Composite CI

on:
  push:
  pull_request:
    branches: [ main ]
  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest
    steps:

>       - uses: actions/checkout@v4
>       - uses: ./.github/actions/setup-and-install
>         with:
>           node-version: '20'
>       - run: npm test
> 
>   lint:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: actions/checkout@v4
>       - uses: ./.github/actions/setup-and-install
>         with:
>           node-version: '20'

Reviewer: ```suggestion
      - uses: actions/checkout@v5
      - uses: ./.github/actions/setup-and-install
        with:
          node-version: '24'
      - run: npm test

  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v5
      - uses: ./.github/actions/setup-and-install
        with:
          node-version: '24'
```

## Part 2: Authoring a JavaScript Action

JavaScript actions run directly on the runner and can use the GitHub Actions Toolkit to speed development. They are fast to execute and easy to distribute via GitHub Marketplace or hosted privately in your enterprise.

In this lab we will build a simple JavaScript action that greets a user by name.

Start by creating a new folder and initializing a base action: `hello-js-action`:

```bash
mkdir hello-js-action && cd hello-js-action
npm init -y
npm install @actions/core @actions/github

# Use rollup for bundling
npm install --save-dev rollup @rollup/plugin-commonjs @rollup/plugin-node-resolve
npm install -g rollup
```

The base of an action is the `action.yml` file which defines inputs, outputs, and the runtime.  Create `hello-js-action/action.yml` with the following content:

```yaml
name: Hello JS Action
description: Greet someone and record the time

inputs:
  who-to-greet: # id of input
    description: Who to greet
    required: true
    default: World

outputs:
  time: # id of output
    description: The time we greeted you

runs:

>   using: node20

Reviewer: ```suggestion
  using: node24
```

  main: dist/index.js
```

Note that this defines inputs required, but also specified the runtime and the main script to run.

Next, create `src/index.js`, the main script for the action:

```javascript
import * as core from "@actions/core";
import * as github from "@actions/github";

try {
  // `who-to-greet` input defined in action metadata file
  const nameToGreet = core.getInput("who-to-greet");
  core.info(`Hello ${nameToGreet}!`);

  // Get the current time and set it as an output variable
  const time = new Date().toTimeString();
  core.setOutput("time", time);

  // Get the JSON webhook payload for the event that triggered the workflow
  const payload = JSON.stringify(github.context.payload, undefined, 2);
  core.info(`The event payload: ${payload}`);
} catch (error) {
  core.setFailed(error.message);
}

```

Finally we can create a workflow to test our action.  In your repo add `.github/workflows/test-js-action.yml`:

```yaml
name: Test JS Action

on:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:

>       - uses: actions/checkout@v4

Reviewer: ```suggestion
      - uses: actions/checkout@v5
```
